Revamp post filtering system

ClearlyClaire opened the pull request:
Fixes #18035

## New filter management UI

It's now possible to regroup filter keywords in categories.

![image](https://user-images.githubusercontent.com/384364/174633036-372736c6-fc94-4475-9bdc-6539e21d5b94.png)
![image](https://user-images.githubusercontent.com/384364/174769862-8dfe3ce3-98ff-428a-93af-9e8694c0c72d.png)

## New filtered post UI

Instead of just displaying “Filtered”, the filter category name is displayed, and an opportunity is given to show the filtered post:

https://user-images.githubusercontent.com/384364/174632923-08f681df-9ed7-41e7-befb-03884dccee63.mp4

## API changes and implementation guidelines

Note that everything described here is subject to change before the feature actually gets released.

### Backward compatibility

This PR provides **limited** backward compatibility with the `/api/v1/filters` API:
- `GET /api/v1/filters` is supported and will return all keywords, including when multiple keywords are in a single category. Existing pre-update keywords should retain their identifier, though there is no guarantee this will be the case.
- `GET /api/v1/filters/:id` is supported and will return a keyword, even if that keyword shares a category with other keywords. If that keyword was created pre-update, it should retain its identifier, though there is no guarantee this will be the case.
- `POST /api/v1/filters` is supported, and will create a one-keyword filter with a title matching the keyword
- `PUT /api/v1/filters/:id` is partially supported, **and will error out if it attempts to change `expires_in`, `irreversible` or `context` for a filter shared by multiple keywords**.

### New API

#### New entities

- `Filter`: has attributes `id`, `title`, `context`, `expires_at`, `filter_action` (currently either `hide` or `warn`), and optionally `keywords` (list of `FilterKeyword` entities).
  This entity will likely be extended in the future with attributes for other rule types (e.g. `conversations`).
- `FilterKeyword`: has attributes `id`, `keyword` and `whole_word`
- `FilterResult`: has attributes `filter` (a `Filter` without a `keywords` attribute) as well as `keyword_matches`.
  This entity will likely be extended in the future with attributes for other rule types (e.g. `conversation_matches`).

#### Changes to other entities

- `Status` entities now optionally have a `filtered` attribute which is a list of `FilterResult` entities.

#### New API endpoints

- `GET /api/v2/filters` to list filters (including keywords)
- `POST /api/v2/filters` to create a new filter
  `keywords_attributes` can also be passed to create keywords in one request
- `GET /api/v2/filters/:id` to read a particular filter
- `PUT /api/v2/filters/:id` to update a filter
  `keywords_attributes` can also be passed to edit, delete or add keywords in one request
- `DELETE /api/v2/filters/:id` to delete a particular filter
- `GET /api/v2/filters/:id/keywords` to list keywords for a filter
- `POST /api/v2/filters/:filter_id/keywords/:id` to add a new keyword to a filter
- `GET /api/v1/filters/keywords/:id` to read a particular keyword
- `PUT /api/v1/filters/keywords/:id` to edit a particular keyword
- `DELETE /api/v1/filters/keywords/:id` to delete a particular keyword

### Implementation guidelines

#### Determining matched filters

Since the server provides a `filtered` attribute on `Status` entities, that can be used to check which filters apply without implementations needing to implement the rule matching themselves.

However, client implementations may still want to perform rule matching client-side, as this would allow retroactively apply filter changes without re-fetching posts from the server. When doing so, they should take care of not ignoring `filtered` entries for which there are other attributes than `keyword_matches`, so as to handle future extensions of the filtering system.

#### Applying actions on matched rules

Matched filters need to be filtered based on context (`home`, `notifications`, `public`, `thread` or `profile`) and expiration date.

When at least one active matched filter has `hide` for `filter_action`, the post should not be shown at all. Otherwise, if at least one active matched filter has `warn` for `filter_action`, the post should be hidden with a warning, and the user should be able to reveal the post after being informed of which filters matched (identified by `title` rather than the exact matched keywords).

For extension purposes, unknown values for `filter_action` should be treated as `warn`.

## TODO

- [x] internally support multiple keywords per filter category
  - [x] add model
  - [x] add migrations
  - [x] update `/filters` controllers and views
  - [x] update REST API controllers
- [x] interface to manage multiple keywords per filter category
- [x] switch from `irreversible` to multiple actions
- [x] write the new API
  - [x] `/api/v2/filters` (list and manage filters)
  - [x] server-side filtering
  - [x] streaming events
- [x] update the Web UI
- [ ] add filter types (those can safely be postponed to another PR)
  - [ ] muted statuses (individual)
  - [ ] muted threads
  - [ ] muted mentions (prerequisite for #17631 imho)

___

This project was funded through the [NGI0 Discovery](https://nlnet.nl/discovery) Fund, a fund established by [NLnet](https://nlnet.nl/) with financial support from the European Commission's [Next Generation Internet](https://ngi.eu/) programme, under the aegis of DG Communications Networks, Content and Technology under grant agreement No 825322.

ClearlyClaire committed:
Add model for custom filter keywords

ClearlyClaire committed:
Use CustomFilterKeyword internally

Does not change the API

ClearlyClaire committed:
Add migration tests

ClearlyClaire committed:
Remove whole_word column from custom_filters (covered by custom_filter_keywords)

ClearlyClaire committed:
Redesign /filters

Instead of a list, present a card that displays more information and handles
multiple keywords per filter.

ClearlyClaire committed:
Fix /filters/edit and /filters/new

ClearlyClaire committed:
Redesign /filters/new and /filters/edit to add and remove keywords

This adds a new gem dependency: cocoon, as well as a npm dependency:
cocoon-js-vanilla. Those are used to easily populate and remove form fields
from the user interface when manipulating multiple keyword filters at once.

ClearlyClaire commented:
It's not possible yet to add multiple keywords to a category (missing API, and the `/settings` page need more work), but here is how it gets displayed:
![image](https://user-images.githubusercontent.com/384364/164510583-efb6f644-dcd1-4921-a017-76ae78de55c4.png)

ariasuni commented:
I like it a lot, but it looks a bit too large in my opinion. I have currently 33 filters, which could probably be changed into 10 groups, and I don’t want to scroll too much.

I feel like keywords should be between title and «metadata» (places where filter is applied, expiration time). Also, the longest metadata line in English will overflow; maybe metadata should be placed on its own line below title/keywords/action buttons so that metadata line never overflows on desktop (and separated into two different lines on mobile?).

ClearlyClaire commented:
> I like it a lot, but it looks a bit too large in my opinion. I have currently 33 filters, which could probably be changed into 10 groups, and I don’t want to scroll too much.

I honestly have the same concern :/

> I feel like keywords should be between title and «metadata» (places where filter is applied, expiration time).

I am not sure about that, I want to eventually have other kinds of filters, e.g. filtering specific conversations (see the following mockup):

![image](https://user-images.githubusercontent.com/384364/164517609-495f224d-d4e4-4671-935e-20220c4c198e.png)

> Also, the longest metadata line in English will overflow; maybe metadata should be placed on its own line below title/keywords/action buttons so that metadata line never overflows on desktop (and separated into two different lines on mobile?).

Also a good point.

ClearlyClaire commented:
@ariasuni reworked a bit that design by moving things around and reducing some margins, I think it addresses most of your concerns

![image](https://user-images.githubusercontent.com/384364/164537507-5c47d578-6c20-44ab-9e2c-7969a6365460.png)

ariasuni commented:
It much better in my opinion! Keyword list and conversation count could be side-to-side, though.

ClearlyClaire committed:
Add /api/v2/filters to edit filter with multiple keywords

Entities:
- `Filter`: `id`, `title`, `filter_action` (either `hide` or `warn`), `context`
  `keywords`
- `FilterKeyword`: `id`, `keyword`, `whole_word`

API endpoits:
- `GET /api/v2/filters` to list filters (including keywords)
- `POST /api/v2/filters` to create a new filter
  `keywords_attributes` can also be passed to create keywords in one request
- `GET /api/v2/filters/:id` to read a particular filter
- `PUT /api/v2/filters/:id` to update a new filter
  `keywords_attributes` can also be passed to edit, delete or add keywords in
   one request
- `DELETE /api/v2/filters/:id` to delete a particular filter
- `GET /api/v2/filters/:id/keywords` to list keywords for a filter
- `POST /api/v2/filters/:filter_id/keywords/:id` to add a new keyword to a
   filter
- `GET /api/v2/filter_keywords/:id` to read a particular keyword
- `PUT /api/v2/filter_keywords/:id` to edit a particular keyword
- `DELETE /api/v2/filter_keywords/:id` to delete a particular keyword

ClearlyClaire committed:
Change from `irreversible` boolean to `action` enum

ClearlyClaire committed:
Remove irrelevent `irreversible_must_be_within_context` check

ClearlyClaire committed:
Fix /filters/new and /filters/edit with update for filter_action

ClearlyClaire committed:
Fix Rubocop/Codeclimate complaining about task names

ClearlyClaire committed:
Refactor FeedManager#phrase_filtered?

This moves regexp building and filter caching to the `CustomFilter` class.

This does not change the functional behavior yet, but this changes how the
cache is built, doing per-custom_filter regexps so that filters can be matched
independently, while still offering caching.

ClearlyClaire committed:
Perform server-side filtering and output result in REST API

ClearlyClaire committed:
Add some tests

ClearlyClaire committed:
Use the new API in the WebUI

- use client-side logic for filters we have fetched rules for.
  This is so that filter changes can be retroactively applied without
  reloading the UI.
- use server-side logic for filters we haven't fetched rules for yet
  (e.g. network error, or initial timeline loading)

ClearlyClaire committed:
Minor optimizations and refactoring

ClearlyClaire committed:
Fix numerous filters_changed events being sent when editing multiple keywords at once

ClearlyClaire committed:
Perform server-side filtering on the streaming server

ClearlyClaire commented:
I have finally added the UI as well as filtering in the streaming server. I am confident that last bit works as intended, but I am not sure it is worth the code duplication and the additional JS dependency. Furthermore, whether this filtering is done or not server-side does not matter for the WebUI, as it performs its own checks client-side (so that changing the filters is retroactive).

The UI for filtered toots look like this:

https://user-images.githubusercontent.com/384364/174632265-d2288055-cb23-4a1c-8622-35fd12c7c71a.mp4

ariasuni commented:
Is «Hide» equivalent to the current «Drop instead of hide», or is it simply the same as «Warn» but without any visual clue?

I suggest «Hide with a warning» and «Hide completely» (or something like that) if that’s the case.

ClearlyClaire commented:
> Is «Hide» equivalent to the current «Drop instead of hide», or is it simply the same as «Warn» but without any visual clue?
> 
> I suggest «Hide with a warning» and «Hide completely» (or something like that) if that’s the case.

It's more or less the case, the “drop instead of hide” is not completely reversible, while this filter type is supposed to be.

ClearlyClaire committed:
Change the wording of filter action labels

ClearlyClaire committed:
Fix issues pointed out by linter

ariasuni reviewed (approved):
Approving the design (didn’t look at the code).

Gargron commented on app/lib/feed_manager.rb:
```diff
@@ -424,24 +424,12 @@ def filter_from_list?(status, list)
   def phrase_filtered?(status, receiver_id, context)
-    active_filters = Rails.cache.fetch("filters:#{receiver_id}") { CustomFilter.where(account_id: receiver_id).active_irreversible.to_a }.to_a
+    active_filters = CustomFilter.cached_filters_for(receiver_id)
+    active_keywords = active_filters.filter_map { |custom_filter, rules| rules[:keywords] if custom_filter.context.include?(context.to_s) && custom_filter.irreversible? }
```
Irreversability of filters still makes me uneasy as this is basically an entirely different code path for filtering stuff to how it's done otherwise. I wonder if it would not be, after all, better to rely entirely on clients hiding items. I know that I initially came up with the `irreversible` flag with the reasoning being to ensure correct pagination (not have to deal with "empty" pages being returned) but I think it might be worth it to have a more straightforward filtering path.

ClearlyClaire replied:
It's actually a leftover from the early rewriting before I had updated the API everywhere. I intend to remove this to not make it “irreversible”.

ClearlyClaire replied:
Removed.

Gargron commented on app/serializers/rest/status_serializer.rb:
```diff
@@ -13,6 +13,7 @@ class REST::StatusSerializer < ActiveModel::Serializer
   attribute :muted, if: :current_user?
   attribute :bookmarked, if: :current_user?
   attribute :pinned, if: :pinnable?
+  has_many :filter_results, serializer: REST::FilterResultSerializer, if: :current_user?
```
How about naming it `filtered`, or `filters`?

ClearlyClaire replied:
Renamed to `filtered`.

Gargron commented on config/routes.rb:
```diff
@@ -588,6 +588,10 @@
+      resources :filters,     only: [:index, :create, :show, :update, :destroy] do
+        resources :keywords, only: [:index, :create], controller: 'filters/keywords'
+      end
+      resources :filter_keywords, only: [:show, :update, :destroy], controller: 'filters/keywords'
```
I'm not entirely sure why this exists at the top level as well as under the filters resource. I would prefer to avoid mixing namespacing schemes (`filter_keywords` and `filters/:id/keywords`).

ClearlyClaire replied:
Those are recommendations from https://guides.rubyonrails.org/routing.html#limits-to-nesting

Gargron replied:
Right, but most Rails apps are simple CRUD and don't really care how URLs look, and aren't stuck with them for years to come. You could try doing something like:

```ruby
namespace :filters do
  resources :keywords, only: [:show, :update, :destroy], controller: 'filters/keywords'
end
```

So the result would be something like:

```
GET  /api/v1/filters/:id/keys
POST /api/v1/filters/:id/keys

GET    /api/v1/filters/keys/:id
PUT    /api/v1/filters/keys/:id
DELETE /api/v1/filters/keys/:id
```

ClearlyClaire replied:
Moved `/api/v2/filter_keywords/` to `/api/v1/filters/keywords/`.

Gargron commented on db/migrate/20220613110711_migrate_custom_filters.rb:
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class MigrateCustomFilters < ActiveRecord::Migration[6.1]
```
Would this not be a post-deployment migration?

ClearlyClaire replied:
I am not sure why it would be. It won't break anything if the running code is not up-to-date, but it will cause issues if admins forget to run it.

Gargron commented on streaming/index.js:
```diff
@@ -512,7 +516,8 @@ const startWorker = async (workerId) => {
   const subscribeHttpToSystemChannel = (req, res) => {
-    const systemChannelId = `timeline:access_token:${req.accessTokenId}`;
+    const accessTokenChannelId = `timeline:access_token:${req.accessTokenId}`;
+    const systemChannelId = `timeline:token:${req.accountId}`;
```
Naming it `token` and then using `accountId` is not very intuitive. I suppose this has got to be something like `timeline:${req.accountId}:system` to be consistent with other channel names (e.g. `timeline:${req.accountId}:notifications`)

ClearlyClaire replied:
That was an error on my part! I intended to name that channel `timeline:system:${req.accountId}`! Fixed it.

Wonder about `timeline:${req.accountId}:system` VS `timeline:system:${req.accountId}` though

Gargron commented on streaming/index.js:
```diff
@@ -674,17 +681,84 @@ const startWorker = async (workerId) => {
           done();
 
-          if (values[0].rows.length > 0 || (values.length > 1 && values[1].rows.length > 0)) {
+          if (values[0].rows.length > 0 || (accountDomain && values[1].rows.length > 0)) {
```
The previous condition ensured there would not be access to an array element that returns undefined. From viewing the diff alone it's not obvious to me that there would always be two rows returned when `accountDomain` is truthy.

ClearlyClaire replied:
There are has many array elements as there are in `queries`. When `accountDomain` is truthy, a specific query is pushed. If `accountDomain` is false-y but the filter query is pushed, `values.length > 1` holds true but `values[1]` is about the filters query, not the domain block one.

Gargron commented on streaming/index.js:
```diff
@@ -674,17 +681,84 @@ const startWorker = async (workerId) => {
+                cache[row.id] = {
+                  keywords: [[row.keyword, row.whole_word]],
+                  expires_at: row.expires_at,
+                  repr: {
```
What does `repr` stand for here?

ClearlyClaire replied:
“representation”. But it can be anything else.

Gargron commented on app/javascript/mastodon/selectors/index.js:
```diff
@@ -62,27 +62,41 @@ const regexFromFilters = filters => {
     const serverSideType = toServerSideType(contextType);
-    const filters = state.get('filters', ImmutableList()).filter(filter => filter.get('context').includes(serverSideType) && (filter.get('expires_at') === null || Date.parse(filter.get('expires_at')) > (new Date())));
+    const now = new Date();
+    const filters = state.get('filters', ImmutableMap()).toList().filter(filter => filter.get('context').includes(serverSideType) && filter.get('keywords') && filter.get('keywords').size > 0 && (filter.get('expires_at') === null || filter.get('expires_at') > now));
```
As I understand, server-side filtering logic is still re-implemented client-side to handle filter expiration, thereby reducing the benefits of the new implementation. Ideally, the client would not need to construct its own regexes at all. Why can't the server-side match in the API response include an `expires_at` property, and then the client can check *that*, but does not have to re-match any filter regexes?

ClearlyClaire replied:
The API response includes the `expires_at`, but the whole client-side reimplementation is not for that purpose. It's for when filters get changed (new filters added require re-filtering everything against the added filters, and keywords changing require re-filtering everything against the updated filters).

ClearlyClaire replied:
Yes, a page refresh is enough. I'll see how much simpler the WebUI code gets by removing this.

ClearlyClaire replied:
I'm not too sure about dropping this behavior, but I pushed a commit anyway, so you can judge whether the simplification is worth it.

Gargron commented:
Design-wise, my only comment so far is that "(show anyway)" (lower case, in paranthesis) has no precedent in our UIs. The more expected way would be to have it as "Show anyway", separated from previous text by a period (see e.g. empty column messages)

ClearlyClaire commented on Gemfile:
```diff
@@ -153,3 +153,5 @@ gem 'concurrent-ruby', require: false
 
 gem 'xorcist', '~> 1.1'
+
+gem 'cocoon', '~> 1.2'
```
That's the Ruby gem. The JS counterpart I use, `cocoon-js-vanilla` is a pure JS reimplementation of the same thing

ClearlyClaire committed:
Change design of “Show anyway” link in accordence to review comments

ClearlyClaire committed:
Drop “irreversible” filtering behavior

ClearlyClaire committed:
Move /api/v2/filter_keywords to /api/v1/filters/keywords

ClearlyClaire commented:
Updated the design according to your feedback:
![image](https://user-images.githubusercontent.com/384364/175887259-92d9d42f-1841-4b71-b984-fe5cad1fd9bf.png)

ClearlyClaire committed:
Rename `filter_results` attribute to `filtered`

ClearlyClaire committed:
Rename REST::LegacyFilterSerializer to REST::V1::FilterSerializer

ClearlyClaire commented on app/controllers/api/v1/filters_controller.rb:
```diff
@@ -8,21 +8,32 @@ class Api::V1::FiltersController < Api::BaseController
 
   def index
-    render json: @filters, each_serializer: REST::FilterSerializer
+    render json: @filters, each_serializer: REST::LegacyFilterSerializer
```
Renamed to `REST::V1::FilterSerializer`

ClearlyClaire committed:
Fix systemChannelId value in streaming server

ClearlyClaire committed:
Simplify code by removing client-side filtering code

The simplifcation comes at a cost though: filters aren't retroactively
applied anymore.

Lukas2112 commented:
Wasn't it also a thought that you can see whose post (user handle) was filtered, @ClearlyClaire ?

ClearlyClaire commented:
Could be done in a follow-up PR but I'm unsure what it should look like (glitch-soc has something similar but it looks pretty rough)

zunda commented:
Merging this, it seems that boosts by followee are unexpectedly filtered out on my HTL. Might there be anything I'd check?

Freeplayg commented:
Yea, I'm having the same issue here... I have no filters set. 
(video is slightly borked but you can still see)

https://user-images.githubusercontent.com/40742947/176452206-a9d9c04d-7160-4b9f-b5ff-b11b069220d3.mp4

ClearlyClaire commented:
Thanks for the report, I will investigate this!

ClearlyClaire commented:
Ok, that was caused by renaming `filter_results` to `filtered`, should be fixed in #18744

zunda commented:
Thank you! I've just confirmed #18744 fixed the problem. Boosts are now shown as before.

mashirozx commented:
Why I went into such kind of diff when doing db migration?
<img width="994" alt="image" src="https://user-images.githubusercontent.com/16148054/177029987-932e8355-f04f-4af2-9974-d60d25d1dbe0.png">

ClearlyClaire commented:
> **Bug**: When creating a new filter i get an error message upon trying to save this including a set keyword. The error does not appear, when not setting a keyword.

What error message do you get? This is likely #18742

> **Unclear**: Is the filter active, even when no keyword is set? It seems so, but that wouldn't make any sense, would it? Generally the keyword is meant to trigger the filter, right?

When no keyword is set, the filter does nothing. There are other filter types planned, but maybe the UI should be tweaked to make it clearer that the filter won't do anything without a keyword.

> **GUI**: I think the new design wastes space and find the former one line display much more usable to get a quick overview on many filters.

I'm split on this one. I can understand how the one-line display may be more usable if you have many different filters, but I think this one is more usable if your keywords are logically grouped, which I think fits the general case more.

jayrope commented:
Thanks for getting into this, ClearlyClaire.

> What error message do you get? This is likely #18742

It is similar, but i get the server error already upon creating a new filter when adding a keyword at all (v. 3.5.3 on mastodon.online that is).
 
> When no keyword is set, the filter does nothing. There are other filter types planned, but maybe the UI should be tweaked to make it clearer that the filter won't do anything without a keyword.

I suppose i cannot currently create _any_ new and working filters then, as setting a keyword produces that error. Uhm.

> > **GUI**: ...
> 
> I'm split on this one. I can understand how the one-line display may be more usable if you have many different filters, but I think this one is more usable if your keywords are logically grouped, which I think fits the general case more.

Why not give the user a choice of display? That way you'll satisfy both power filterers as well as people, who use only a few. I am filtering a lot. It is one of my most used tools to regulate the flood into a lean experience. The current display turns out to be a major scroll monster for me :).

Thank you again for anything you do.

ClearlyClaire commented:
> Thanks for getting into this, ClearlyClaire.
> 
> > What error message do you get? This is likely #18742
> 
> It is similar, but i get the server error already upon creating a new filter when adding a keyword at all (v. 3.5.3 on mastodon.online that is).

Yeah, it's likely the same thing then. It can occur whenever adding a keyword, whether there are keywords in the filter or not.

> > When no keyword is set, the filter does nothing. There are other filter types planned, but maybe the UI should be tweaked to make it clearer that the filter won't do anything without a keyword.
> 
> I suppose i cannot currently create _any_ new and working filters then, as setting a keyword produces that error. Uhm.

Yes, unfortunately, this issue may prevent you from creating any new filter… I think you can retry over and over again and eventually it should work… but you may need to retry a lot of times :/

> > > **GUI**: ...
> > 
> > 
> > I'm split on this one. I can understand how the one-line display may be more usable if you have many different filters, but I think this one is more usable if your keywords are logically grouped, which I think fits the general case more.
> 
> Why not give the user a choice of display? That way you'll satisfy both power filterers as well as people, who use only a few. I am filtering a lot. It is one of my most used tools to regulate the flood into a lean experience. The current display turns out to be a major scroll monster for me :).

Because it's not just the display that changed, but the whole model, and the old display would not quite work with the new model…

ariasuni commented:
I have the new filter system and my old filters are still here.

ClearlyClaire commented:
> Okay, that sounds good, are they then "uncategorised" or how are the old filters ordered?

They are all added as their own category.

stom79 commented:
Hi @ClearlyClaire 
I try to implement that feature with the API. Could you tell more about payload for POST/PUT endpoints?
For instance with `/api/v2/filters` 

I tried this:

```
    @FormUrlEncoded
    @POST("filters")
    Call<Filter> addFilter(
            @Header("Authorization") String token,
            @Field("title") String title,
            @Field("expires_at") Date expires_at,
            @Field("filter_action") String filter_action,
            @Field("context[]") List<String> context,
            @Field("keywords_attributes[]") List<Filter.KeywordsAttributes> keywordsAttributes
    );
```
```java
    public static class KeywordsAttributes implements Serializable {
        @SerializedName("id")
        public String id;
        @SerializedName("keyword")
        public String keyword;
        @SerializedName("whole_word")
        public boolean whole_word;
        @SerializedName("_destroy")
        public boolean _destroy;
    }
```

But Keywords are not added

If I use the `PUT` call, only context is changed (title, filter_action, keywords) are not updated.

ClearlyClaire commented:
That is weird, those should work… or at least I can't see anything immediately wrong with it. `POST` should not include `id`s for keyword attributes though.

Are you sure you are hitting `/api/v2/filters` and not `/api/v1/filters` still?

trwnh commented:
@stom79 unsure if related, but `expires_at` should be `expires_in`

i can't replicate any issues with a raw curl `POST http://mastodon.local/api/v2/filters?title=test&context[]=home&keywords_attributes[][keyword]=word` though...

stom79 commented:
yes I use `/api/v2` with `/filters`, I can make it works by using 

```java
  @Field("keywords_attributes[][id]") List<String> keywordId,
  @Field("keywords_attributes[][keyword]") List<String> keywords,
  @Field("keywords_attributes[][whole_word]") List<Boolean> wholeWords
```

instead of 

```java
 @Field("keywords_attributes[]") List<Filter.KeywordsAttributes> keywordsAttributes
 ```

But `whole_word` is always set to true. I will check why.

`PUT` to  `/api/v2/filters`  should also update keywords if the id is passed? Currently that doesn't add new ones but it doesn't change their value.

Tak commented:
@stom79 I'm seeing a similar issue here, also using retrofit, can you clarify what ended up working for you?
If I `PUT` to `/api/v2/filters/{id}` with `@Field("keywords_attributes[][id]") List<String> keywordId` etc., it creates *new* keywords with new ids (yes, I'm passing the old ids) (and gives 422 if I pass `whole_word` at all)
If I use a list of objects or maps instead, the call succeeds but the keywords are ignored.

stom79 commented:
@Tak Our endpoints are here for filters 

https://codeberg.org/tom79/Fedilab/src/branch/main/app/src/main/java/app/fedilab/android/client/endpoints/MastodonFiltersService.java

And our activity that manages filters

https://codeberg.org/tom79/Fedilab/src/branch/main/app/src/main/java/app/fedilab/android/activities/FilterActivity.java

We created objects for retrofit 

https://codeberg.org/tom79/Fedilab/src/branch/main/app/src/main/java/app/fedilab/android/client/entities/api/Filter.java#L74

And 
https://codeberg.org/tom79/Fedilab/src/branch/main/app/src/main/java/app/fedilab/android/client/entities/api/Filter.java#L89

We use them directly with retrofit for POST/PUT calls 
https://codeberg.org/tom79/Fedilab/src/branch/main/app/src/main/java/app/fedilab/android/client/endpoints/MastodonFiltersService.java#L81

Currently, I am not with my computer, but I will check that more carefully tomorrow.
